### ISSUE TITLE
Improve efficiency of func trainOnBatch

### ISSUE THREAD

**Issue by XinDongol** [mariob6/style_text#22]
In function `trainOnBatch`
https://github.com/mberaha/style_text/blob/b5598f69334d8cfe04856ae1c6af7a10d916da90/src/style_transfer.py#L365

`_runBatch` is called for three times to compute losses for `discriminator1`, `discriminator2`, and `autoencoder` as follows.
https://github.com/mberaha/style_text/blob/b5598f69334d8cfe04856ae1c6af7a10d916da90/src/style_transfer.py#L372-L374

https://github.com/mberaha/style_text/blob/b5598f69334d8cfe04856ae1c6af7a10d916da90/src/style_transfer.py#L382-L384

https://github.com/mberaha/style_text/blob/b5598f69334d8cfe04856ae1c6af7a10d916da90/src/style_transfer.py#L392-L394

**I was wondering whether I can compute all losses in one forward pass.**

**Comment by mberaha** [comment]
Hey there! thanks for this. I must say, this was my first ever project in pytorch, so no guarantees on it :)

You could split `_runBatch()` in two functions, the first one computes all the parameters needed for the losses, like `h_teacher, h_professor, ...`, and the second one instead computes the right losses and backpropagates.
This would allow to call the first function only once. However (and I have not tried it) you should also take care of the effect of zeroing the gradients, as you might lose some information along the way.

I am not going to be able to work on this any time soon, so I am closing this unless you want to propose a PR :)

**Comment by XinDongol** [comment]
Thanks for your reply. I modified the code according to your helpful suggestions and it worked well.
